feat(conductor, relayer)!: batch multiple Sequencer blocks to save on Celestia fees

SuperFluffy opened the pull request:
## Summary
Batch mutiple Sequencer blocks into single Celestia blobs to reduce fee payments.

## Background
Until now, each Sequencer block was turned into multiple blobs (one for overall sequencer block metadata, and one blob per rollup that had transactiosn in the sequencer block). This wasn't as efficient as it could be because the new compression scheme introduced in https://github.com/astriaorg/astria/pull/1006 can only come to bear with more bytes to compress.

Relayer will collect sequencer blocks up to a total (compressed) size of 1MB (1/2 of the current max of 2MB that Celestia blocks can be).

## Changes
- Introduce protobuf messages `astria.sequencerblock.v1alpha1.CelestiaHeaderList` and `astria.sequencerblock.v1alpha1.CelestiaRollupDataList`
- Rename `astria.sequencerblock.v1alpha1.CelestiaRollupBlob` to `astria.sequencerblock.v1alpha1.CelestiaRollupData`
- Rename `astria.sequencerblock.v1alpha1.CelestiaSequencerBlob` to `astria.sequencerblock.v1alpha1.CelestiaHeader`
- Collect Sequencer Blocks into the `*List` protobuf messages before posting them to Celestia (instead of splitting up each Sequencer block into mutiple blobs and posting them one by one).

## Testing
Add unit tests around the next submission aggregation logic. Update conductor blackbox tests.

## Metrics
+ `CELESTIA_PAYLOAD_CREATION_LATENCY`: histogram with microsecond units to track the time it takes to create a payload of Celestia blobs (encoding + compressing all protobufs)
+ metrics for reporting compression ratio and total compressed payload size were moved from the payload/blob construction phase to the submission phase.

## Breaking Changelist
- Relayer and Conductor write/read new protobuf messages to/from Celestia. 

## Related Issues
Closes https://github.com/astriaorg/astria/issues/1042
Closes https://github.com/astriaorg/astria/issues/1049

Fraser999 commented on crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:
```diff
@@ -1,142 +1,610 @@
-fn serialize_height<S>(height: &SequencerHeight, serializer: S) -> Result<S::Ok, S::Error>
+use crate::IncludeRollup;
+
+/// The maximum permitted payload size in bytes that relayer will send to Celestia.
+///
+/// Taken as half the maximum block size that Celestia currently allows (2 MB at the moment).
+const MAX_PAYLOAD_SIZE_BYTES: usize = 1_000_000;
+
+/// The maximum size that a payload can have with a single sequencer block included.
+const MAX_EXCEPTIONALLY_LARGE_PAYLOAD_BYTES: usize = 1_500_000;
```
We could maybe have a const for max Celestia size and derive these two consts from that one?  Should make it slighter harder to get wrong if/when the max size specified by Celestia changes.

SuperFluffy replied:
Removed this as part of d913803

Fraser999 commented on crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:
```diff
@@ -1,142 +1,610 @@
+        let mut rng = ChaChaRng::seed_from_u64(0);
+        let mut next_submission = NextSubmission::new(include_all_rollups());
+
+        // using the upper limit defined in the constant and add 100KB of extra bytes to ensure
```
```suggestion
        // using the upper limit defined in the constant and add 1000 extra bytes to ensure
```

SuperFluffy replied:
Fixed in d913803

joroshiba commented on proto/sequencerblockapis/astria/sequencerblock/v1alpha1/celestia.proto:
I haven't ever loved the naming of these, and I think the new names make it if anything worse.

The core issue is that these are `Celestia` prefixed but the data has nothing to do with celestia. Especially the new `CelestiaHeader` is not great because Celestia has headers and have nothing to do with these.

I propose we get rid of the `Celestia` prefix entirely and have `RollupData` and `SequencerBlockMetadata`, and their list forms. The Sequencer Block one is hard because it's the `SequencerBlockHeader` plus some information. I think metadata encompasses this alright.

SuperFluffy replied:
Unfortunately `RollupData` was already taken in the same protobuf package. I chosen the following nomenclature instead:
+ `SubmittedMetadata`: since the full type name is actually `astria.sequencerblock.v1alpha1.SubmittedMetadata` it's clear that this pertains to a `sequencerblock`.
+ `SubmittedRollupData`

joroshiba commented on crates/astria-conductor/src/celestia/convert.rs:
```diff
@@ -64,42 +72,73 @@ pub(super) fn decode_raw_blobs(
+        }
+    }
+
+    fn extend_from_rollup_data_list_if_welformed(&mut self, list: CelestiaRollupDataList) {
```
Maybe if we have it return a result just `extend_from_rollup_data_list`?

SuperFluffy replied:
addressed in 5fbcefae

joroshiba commented on crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:
```diff
@@ -1,142 +1,610 @@
+const MAX_PAYLOAD_SIZE_BYTES: usize = 1_000_000;
+
+/// The maximum size that a payload can have with a single sequencer block included.
+const MAX_EXCEPTIONALLY_LARGE_PAYLOAD_BYTES: usize = 1_500_000;
```
I don't think we need a separate restriction size here, although it would be good to know if we end up in this scenario we limit the number of sequenced bytes per sequencer block to 200kb, so hitting even the `1_000_000` limit would be indicative of something going wrong. If anything we should have this be lower than the total size to flag early we are seeing issues.

SuperFluffy replied:
Yes, good point. I have removed this extra limit in d913803

joroshiba commented on crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:
```diff
@@ -1,142 +1,610 @@
+    pub(super) fn try_add(&mut self, block: SequencerBlock) -> Result<(), TryAddError> {
+        let mut input_candidate = self.input.clone();
+        input_candidate.extend_from_sequencer_block(block.clone(), &self.rollup_filter);
+        let payload_candidate = input_candidate.clone().try_into_payload()?;
```
Related: we should have some metrics around time spent compressing

SuperFluffy replied:
Added a metric to track the entire time spent creating the payload, so:
1. encoding all protobufs (the metadata list and lists for all rollups)
2. compressing all encoded protobufs
3. creating the blobs
22c71dea

SuperFluffy commented on crates/astria-conductor/src/celestia/convert.rs:
```diff
@@ -64,42 +72,73 @@ pub(super) fn decode_raw_blobs(
+            self.push_header(header);
+            Ok::<(), CelestiaSequencerBlobError>(())
+        }) {
+            info!(
```
We are actually doing that! `FetchConvertVerifyAndReconstruct::execute` has the `celestia_height` in its span, and we are then decoding the blobs using this:
```rust
        let decoded_blobs = spawn_blocking(move || {
            decode_span
                .in_scope(|| decode_raw_blobs(new_blobs, rollup_namespace, sequencer_namespace))
        })
        .await
```
Hence all events are associated with the correct celestia height.

`decode_raw_blobs` itself has no knowledge of which Celestia height it is operating under - and both `decode_raw_blobs` and the `extend_from_*_if_well_formed` methods are deliberately not returning errors because we expect that blobs are being posted under our namespaces (no way to prevent that).

SuperFluffy commented on crates/astria-conductor/src/celestia/convert.rs:
```diff
@@ -64,42 +72,73 @@ pub(super) fn decode_raw_blobs(
+            self.push_rollup_data(entry);
+            Ok::<(), CelestiaRollupBlobError>(())
+        }) {
+            info!(
```
Answered above, but copying here for context:
> We are actually doing that! `FetchConvertVerifyAndReconstruct::execute` has the `celestia_height` in its span, and we are then decoding the blobs using this:
> 
> ```rust
>         let decoded_blobs = spawn_blocking(move || {
>             decode_span
>                 .in_scope(|| decode_raw_blobs(new_blobs, rollup_namespace, sequencer_namespace))
>         })
>         .await
> ```
> 
> Hence all events are associated with the correct celestia height.
> 
> `decode_raw_blobs` itself has no knowledge of which Celestia height it is operating under - and both `decode_raw_blobs` and the `extend_from_*_if_well_formed` methods are deliberately not returning errors because we expect that blobs are being posted under our namespaces (no way to prevent that).

SuperFluffy commented on crates/astria-conductor/src/celestia/convert.rs:
```diff
@@ -64,42 +72,73 @@ pub(super) fn decode_raw_blobs(
 /// An unsorted [`CelestiaSequencerBlob`] and [`CelestiaRollupBlob`].
```
Addressed as part of 401233a7

SuperFluffy commented on crates/astria-core/src/sequencerblock/v1alpha1/celestia.rs:
Done as part of 401233a7

SuperFluffy commented on crates/astria-sequencer-relayer/src/relayer/write/mod.rs:
```diff
@@ -327,60 +225,92 @@ impl BlobSubmitter {
     let start = std::time::Instant::now();
 
+    debug!(
+        // sequencer_height = %sequencer_height,
```
Yeah, copied it around. I absorbed the fields of this `debug!` event into the `info!` field right above (which is emitted when a new submission is started). Addressed in  82ce52ab

SuperFluffy commented on crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:
```diff
@@ -1,142 +1,610 @@
+    },
+    #[error(
+        "there was no sequencer namespace present in the input so a payload of Celestia blobs \
+         could not be construct"
```
Thank you for noticing this! Fixed in 82ce52ab

SuperFluffy commented on crates/astria-sequencer-relayer/src/relayer/write/conversion.rs:
```diff
@@ -1,142 +1,610 @@
+
+    #[test]
+    fn adding_block_to_full_submission_gets_rejected() {
+        // this tests makes use of the fact that random data is essentialy incompressible so
```
Thanks! 93883b2

SuperFluffy commented:
@joroshiba I have added a histogram metric `CELESTIA_PAYLOAD_CREATION_LATENCY` and changed the protobufs to `SubmittedMetadata`, `SubmittedMetadataList`, `SubmittedRollupData`, `SubmittedRollupDataList`.
